### PR TITLE
Install parsers with `repo.url` as a local path

### PR DESCRIPTION
**Pull request by theHamsta**
This is mainly thought as an internal feature for parser developers to install them from a local path

E.g. when you are working on a Lisp parser:

```
  require "nvim-treesitter.parsers".get_parser_configs().lisp = {
    install_info = {
      url = "~/projects/tree-sitter-lisp",
      files = {"src/parser.c"}
    }
  }
```
and then `:TSInstall lisp`/`:TSUpdate lisp` to compile and install the parser and `:TSInstallFromGrammar lisp` to run `tree-sitter generate` before compiling.

Depends on #503 (only because I edited the same code region).

### CODE REVIEW

**Comment by theHamsta** [comment]
Would be even better if we would `uv_close` the parser shared object and reload on update.

File: lua/nvim-treesitter/install.lua [hunk 1]
```diff
@@ -101,16 +101,30 @@ local function iter_cmd_sync(cmd_list)
...
-local function run_install(cache_folder, install_folder, lang, repo, with_sync)
+local function run_install(cache_folder, install_folder, lang, repo, with_sync, generate_from_grammar)
   parsers.reset_cache()
 
   local path_sep = utils.get_path_sep()
 
   local project_name = 'tree-sitter-'..lang
+  local from_local_path = vim.fn.isdirectory(vim.fn.expand(repo.url)) == 1
+  if from_local_path then
+    repo.url = vim.fn.expand(repo.url)
```
Comment (vigoux) [inline lua/nvim-treesitter/install.lua:112]: Could you avoid calling `expand` twice here ?

File: lua/nvim-treesitter/install.lua [hunk 2]
```diff
@@ -121,8 +135,23 @@ local function run_install(cache_folder, install_folder, lang, repo, with_sync)
...
-  vim.list_extend(command_list, shell.select_download_commands(repo, project_name, cache_folder, revision))
+  if not from_local_path then
+    vim.list_extend(command_list, { shell.select_install_rm_cmd(cache_folder, project_name) })
+    vim.list_extend(command_list, shell.select_download_commands(repo, project_name, cache_folder, revision))
+  end
+  if generate_from_grammar then
+    vim.list_extend(command_list, {
+      {
+        cmd = 'tree-sitter',
+        info = 'Generating source files from grammer.js...',
```
Comment (vigoux) [inline lua/nvim-treesitter/install.lua:146]:
```suggestion
        info = 'Generating source files from grammar.js...',
```

File: lua/nvim-treesitter/install.lua (hunk 1 above)
Comment (theHamsta) [reply lua/nvim-treesitter/install.lua]: fixed

File: lua/nvim-treesitter/install.lua (hunk 2 above)
Comment (theHamsta) [reply lua/nvim-treesitter/install.lua:147]: fixed

**Comment by theHamsta** [comment]
An upgrade for this PR would be to offer the installation of tree-sitter-cli if not installed. But this would be another PR.

**Review by vigoux — APPROVED** [review]

**Pull request merged**